File: XeeSDK/Classes/XeeConnectManager.m
/*
 * Copyright 2016 Eliocity
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#import "XeeConnectManager.h"
#import "Xee.h"

@interface XeeConnectManager () {
    UIWebView *embeddedWV;
    XeeHTTPClient *client;
}

@end

@implementation XeeConnectManager

- (instancetype)init
{
    self = [super init];
    if (self) {
        [self createWebView];
        client = [[XeeHTTPClient alloc] init];
        NSDictionary *json = [[NSUserDefaults standardUserDefaults] objectForKey:@"XeeSDKInternalAccessToken"];
        if(json)
            _accessToken = [[XeeAccessToken alloc] initWithJSON:json];
    }
    return self;
}

-(void)createWebView {
    embeddedWV = [[UIWebView alloc] init];
}

-(void)showWebViewInView:(UIView *)view WithURL:(NSURL *)url {
    CGRect frame = view.bounds;
    frame.origin.y = frame.size.height;
    embeddedWV.frame = frame;
    frame.origin.y = 0;
    [embeddedWV loadRequest:[NSURLRequest requestWithURL:url]];
    UIButton *cancelButton = [[UIButton alloc] initWithFrame:CGRectMake(8.0, 28.0, frame.size.width / 4.0, 40.0)];
    [cancelButton setTitle:NSLocalizedString(@"cancel", nil) forState:UIControlStateNormal];
    [cancelButton setTitleColor:[UIColor colorWithRed:71.0/255.0 green:71.0/255.0 blue:71.0/255.0 alpha:1.0] forState:UIControlStateNormal];
    [cancelButton setBackgroundColor:[UIColor whiteColor]];
    [cancelButton addTarget:self action:@selector(cancel) forControlEvents:UIControlEventTouchUpInside];
    [embeddedWV addSubview:cancelButton];
    [view addSubview:embeddedWV];
    [UIView animateWithDuration:0.4 delay:0.0 options:UIViewAnimationOptionCurveEaseOut animations:^{
        embeddedWV.frame = frame;
    } completion:nil];
}

-(void)cancel {
    CGRect frame = embeddedWV.frame;
    frame.origin.y = frame.size.height;
    [UIView animateWithDuration:0.4 delay:0.2 options:UIViewAnimationOptionCurveEaseOut animations:^{
        embeddedWV.frame = frame;
    } completion:^(BOOL finished) {
        [embeddedWV removeFromSuperview];
        [self.delegate connectManagerDidCancel:self];
    }];
}

-(void)createAccount {
    [self showRegisterPage];
}

-(void)connect {
    // if there's an access token
    if(_accessToken) {
        [Xee log:@"token already exists, try to refresh it"];
        // try to refresh the access token
        [self refreshToken:^(XeeAccessToken *accessToken, NSArray<XeeError *> *errors) {
            if(!errors) {
                [Xee log:@"new token is %@", accessToken];
                // got a new access token, call the handler directly
                if(accessToken)
                    [self.delegate connectManager:self didSuccess:accessToken];
                // else, show the auth page in safari
                else {
                    [self showAuthPage];
                }
            } else {
                [Xee log:@"error refreshing the token"];
                [self.delegate connectManager:self didFailWithErrors:errors];
            }
        }];
    // else show the auth page in safari
    } else {
        [self showAuthPage];
    }
}

-(void)showRegisterPage {
    [Xee log:@"show register page"];
    
    NSString *urlEncodedScopes = [client createURLEncodedStringWithArray:_config.scopes];
    NSURL *url = [NSURL URLWithString:[NSString stringWithFormat:@"%@auth/register?client_id=%@&redirect_uri=%@&scope=%@", self.config.environmentURLString, _config.clientID, _config.redirectURI, urlEncodedScopes]];
    
    UIView *view = [self.delegate connectManagerViewForSignUp];
    [self showWebViewInView:view WithURL:url];
}

-(void)showAuthPage {
    [Xee log:@"show auth page"];
    
    NSString *urlEncodedScopes = [client createURLEncodedStringWithArray:_config.scopes];
    NSURL *url = [NSURL URLWithString:[NSString stringWithFormat:@"%@auth/auth?client_id=%@&redirect_uri=%@&scope=%@", self.config.environmentURLString, _config.clientID, _config.redirectURI, urlEncodedScopes]];
    
    UIView *view = [self.delegate connectManagerViewForLogin];
    [self showWebViewInView:view WithURL:url];
}

-(void)openURL:(NSURL*)url {
    CGRect frame = embeddedWV.frame;
    frame.origin.y = frame.size.height;
    [UIView animateWithDuration:0.4 delay:0.2 options:UIViewAnimationOptionCurveEaseOut animations:^{
        embeddedWV.frame = frame;
    } completion:^(BOOL finished) {
        [embeddedWV removeFromSuperview];
        // gete the code from the redirect URI
        NSString *code = [url.absoluteString componentsSeparatedByString:@"="][1];
        if(![code isEqualToString:@"access_denied"]) {
            [self getToken:code];
        } else {
            [self.delegate connectManager:self didFailWithErrors:nil];
        }
    }];
}

-(void)getToken:(NSString*)code {
    NSData *credentialData = [[NSString stringWithFormat:@"%@:%@", _config.clientID, _config.secretKey] dataUsingEncoding:NSUTF8StringEncoding];
    NSString *base64Credentials = @"Basic ";
    base64Credentials = [base64Credentials stringByAppendingString:[credentialData base64EncodedStringWithOptions:0]];
    NSString *urlString = @"auth/access_token";
    NSDictionary *params = @{@"grant_type":@"authorization_code", @"code":code};
    NSDictionary *headers = @{@"Authorization":base64Credentials,@"Content-Type":@"application/x-www-form-urlencoded"};
    [[client method:@"POST" urlString:urlString params:params headers:headers completionHandler:^(NSData *data, NSArray<XeeError *> *errors) {
        _accessToken = nil;
        if(errors.count == 0) {
            NSDictionary *json = [NSJSONSerialization JSONObjectWithData:data options:NSJSONReadingMutableLeaves error:nil]
            ;
            _accessToken = [[XeeAccessToken alloc] initWithJSON:json];
            [[NSUserDefaults standardUserDefaults] setObject:json forKey:@"XeeSDKInternalAccessToken"];
            [[NSUserDefaults standardUserDefaults] synchronize];
            [Xee log:@"token saved"];
            [self.delegate connectManager:self didSuccess:_accessToken];
        } else {
            [self.delegate connectManager:self didFailWithErrors:errors];
        }
    }] resume];
}

-(void)refreshToken:(void (^)(XeeAccessToken *accessToken, NSArray<XeeError *> *errors))completionHandler {
    NSData *credentialData = [[NSString stringWithFormat:@"%@:%@", _config.clientID, _config.secretKey] dataUsingEncoding:NSUTF8StringEncoding];
    NSString *base64Credentials = @"Basic ";
    base64Credentials = [base64Credentials stringByAppendingString:[credentialData base64EncodedStringWithOptions:0]];
    NSString *urlString = @"auth/access_token";
    NSDictionary *params = @{@"grant_type":@"refresh_token", @"refresh_token":_accessToken.refresh_token};
    NSDictionary *headers = @{@"Authorization":base64Credentials,@"Content-Type":@"application/x-www-form-urlencoded"};
    [[client method:@"POST" urlString:urlString params:params headers:headers completionHandler:^(NSData *data, NSArray<XeeError *> *errors) {
        _accessToken = nil;
        if(errors.count == 0) {
            NSDictionary *json = [NSJSONSerialization JSONObjectWithData:data options:NSJSONReadingMutableLeaves error:nil]
            ;
            _accessToken = [[XeeAccessToken alloc] initWithJSON:json];
            [[NSUserDefaults standardUserDefaults] setObject:json forKey:@"XeeSDKInternalAccessToken"];
            [[NSUserDefaults standardUserDefaults] synchronize];
            [Xee log:@"token saved"];
            completionHandler(_accessToken, nil);
        } else {
            _accessToken = nil;
            [[NSUserDefaults standardUserDefaults] setObject:nil forKey:@"XeeSDKInternalAccessToken"];
            [[NSUserDefaults standardUserDefaults] synchronize];
            [Xee log:@"token cleared"];
            completionHandler(nil, errors);
        }
    }] resume];
}

-(NSData*)createBodyWithParams:(NSDictionary*)params {
    NSString *data = @"";
    for(NSString *key in params) {
        data = [data stringByAppendingString:[NSString stringWithFormat:@"%@=%@&", key, [params objectForKey:key]]];
    }
    return [data dataUsingEncoding:NSUTF8StringEncoding];
}

-(void)disconnect {
    _accessToken = nil;
    [[NSUserDefaults standardUserDefaults] setObject:nil forKey:@"XeeSDKInternalAccessToken"];
    [[NSUserDefaults standardUserDefaults] synchronize];
}

@end

File: XeeSDK/Classes/XeeConnectManager.swift
//
//  XeeConnectManager.swift
//  XeeSDK
//
//  Created by Jean-Baptiste Dujardin on 03/10/2017.
//

import Foundation
import UIKit
import WebKit
import Alamofire

public protocol XeeConnectManagerDelegate: class {
    func didSuccess(token: XeeToken)
    func didFail(WithError error: Error)
    func didCancel()
    func didDisconnected()
}

public class XeeConnectManager: NSObject, WKNavigationDelegate {
    
    public override init() {}
    public static let shared = XeeConnectManager()
    public weak var delegate: XeeConnectManagerDelegate?
    
    public var config: XeeConfig?
    var embeddedWebView: WKWebView!
    var webViewSpinner: UIActivityIndicatorView?
    
    public var token: XeeToken? {
        if let tokenJSON = UserDefaults.standard.object(forKey: "XeeSDKInternalAccessToken") as? [String:Any] {
            let tokenObject: XeeToken = XeeToken(JSON: tokenJSON)!
            return tokenObject
        }else {
            return nil
        }
    }

    public var baseApiURL: String? {
        return config?.baseApiURL
    }
    
    public var baseURL: URL? {
        if let host = config?.environment?.rawValue {
            return URL(string: host)
        }else {
            return nil
        }
    }
    
    public func openURL(URL url: URL) {
        let code = getQueryStringParameter(url: url.absoluteString, param: "code")
        if let code = code {
            XeeRequestManager.shared.getToken(withCode: code, completionHandler: { (error, token) in
                self.embeddedWebView.removeFromSuperview()
                if let token = token {
                    self.delegate?.didSuccess(token: token)
                }else if let error = error {
                    self.delegate?.didFail(WithError: error)
                }
            })
        }
    }
    
    func getQueryStringParameter(url: String, param: String) -> String? {
        guard let url = URLComponents(string: url) else { return nil }
        return url.queryItems?.first(where: { $0.name == param })?.value
    }
    
    func showWebView(WithURLRequest urlRequest: URLRequest) {
        self.embeddedWebView = WKWebView(frame: UIScreen.main.bounds)
        self.embeddedWebView.scrollView.bounces = false
        self.embeddedWebView.navigationDelegate = self
        if self.webViewSpinner == nil {
            self.webViewSpinner = UIActivityIndicatorView(activityIndicatorStyle: .whiteLarge)
            self.webViewSpinner!.color = UIColor.black
            self.webViewSpinner!.center = self.embeddedWebView.center
        }
        self.webViewSpinner!.startAnimating()
        self.embeddedWebView.addSubview(self.webViewSpinner!)
        self.embeddedWebView.load(urlRequest)
        let cancelButton: UIButton = UIButton(frame: CGRect(x: UIScreen.main.bounds.width - UIScreen.main.bounds.width / 4 - 8, y: 28, width: UIScreen.main.bounds.width / 4, height: 40))
        cancelButton.setTitle(NSLocalizedString("cancel", comment: ""), for: .normal)
        cancelButton.setTitleColor(UIColor.init(red: 71.0/255.0, green: 71.0/255.0, blue: 71.0/255.0, alpha: 1.0) , for: .normal)
        cancelButton.backgroundColor = UIColor.white
        cancelButton.addTarget(self, action: #selector(cancel), for: .touchUpInside)
        self.embeddedWebView.addSubview(cancelButton)
        UIApplication.shared.windows.first!.addSubview(self.embeddedWebView)
    }
    
    @objc func cancel() {
        self.embeddedWebView.removeFromSuperview()
        self.delegate?.didCancel()
    }
    
    public func createAccount() {
        showRegisterPage()
    }
    
    public func connect() {
        if token != nil {
            XeeRequestManager.shared.refreshToken(completionHandler: { (error, token) in
                if let error = error {
                    self.delegate?.didFail(WithError: error)
                }else {
                    if let token = token {
                        self.delegate?.didSuccess(token: token)
                    }else {
                        self.showAuthPage()
                    }
                }
            })
        }else {
            showAuthPage()
        }
    }
    
    public func disconnect() {
        XeeRequestManager.shared.revokeToken(completionHandler: nil)
        UserDefaults.standard.set(nil, forKey: "XeeSDKInternalAccessToken")
        UserDefaults.standard.synchronize()
        delegate?.didDisconnected()
    }
    
    func showRegisterPage() {
        
        var parameters: Parameters = [:]
        
        if let clientID = config?.clientID {
            parameters["client_id"] = clientID
        }
        if let redirectURI = config?.redirectURI {
            parameters["redirect_uri"] = redirectURI
        }
        if let host = baseURL, let scopes = config?.scopes {
            let url = URL(string: "\(host)oauth/register?scope=\(scopesURLString(WithScopes: scopes))")!
            if let urlRequest = Alamofire.request(url, parameters: parameters).request {
                showWebView(WithURLRequest: urlRequest)
            }
        }
        
    }
    
    func showAuthPage() {
        var parameters: Parameters = ["response_type":"code"]
        
        if let clientID = config?.clientID {
            parameters["client_id"] = clientID
        }
        if let redirectURI = config?.redirectURI {
            parameters["redirect_uri"] = redirectURI
        }
        if let host = baseURL, let scopes = config?.scopes {
            let url = URL(string: "\(host)oauth/authorize?scope=\(scopesURLString(WithScopes: scopes))")!
            if let urlRequest = Alamofire.request(url, parameters: parameters).request {
                showWebView(WithURLRequest: urlRequest)
            }
        }
    }
    
    func scopesURLString(WithScopes scopes: [String]) -> String {
        var scopesURL = ""
        var i: Int = 0
        for scope in scopes {
            scopesURL.append("\(scope)" + "\(i == scopes.count - 1 ? "" : "+")")
            i = i + 1
        }
        return scopesURL
    }
    
    // Hide spinner on page loaded
    public func webView(_ webView: WKWebView, didFinish navigation: WKNavigation!) {
        if let spinner = self.webViewSpinner {
            spinner.stopAnimating()
            spinner.removeFromSuperview()
        }
    }
    /*
    public func webView(_ webView: WKWebView, decidePolicyFor navigationAction: WKNavigationAction, decisionHandler: @escaping (WKNavigationActionPolicy) -> Void) {
        print(navigationAction.request.allHTTPHeaderFields)
        NSLog("Navigating: " + (navigationAction.request.url?.absoluteString ?? ""))
        if navigationAction.request.url != nil
        {
            let url = navigationAction.request.url
            NSLog("Host: " + (url?.host)!)
            print(navigationAction.request)
            print(url)
            print(navigationAction)
            NSLog("Path: " + url!.path)
            
            if url?.host == "app" {
                
            }
            
            if navigationAction.navigationType == .linkActivated {
                guard let url = navigationAction.request.url else {return}
                webView.load(URLRequest(url: url))
            }
        }
        
        // Default: allow navigation
        decisionHandler(.allow)
    }*/
    
    public func webView(_ webView: WKWebView,
                 decidePolicyFor navigationAction: WKNavigationAction,
                 decisionHandler: @escaping (WKNavigationActionPolicy) -> Void) {

        // if the request is a non-http(s) schema, then have the UIApplication handle
        // opening the request
        if let url = navigationAction.request.url,
            !url.absoluteString.hasPrefix("http://"),
            !url.absoluteString.hasPrefix("https://"),
            UIApplication.shared.canOpenURL(url) {

            // have UIApplication handle the url (sms:, tel:, mailto:, ...)
            UIApplication.shared.open(url, options: [:], completionHandler: nil)

            // cancel the request (handled by UIApplication)
            decisionHandler(.cancel)
        }
        else {
            // allow the request
            decisionHandler(.allow)
        }
    }
}
